src/main/java/com/rapid7/client/dcerpc/msrrp/objects/RPCSecurityDescriptor.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * * Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ */
+package com.rapid7.client.dcerpc.msrrp.objects;
+
+import java.io.IOException;
+import com.rapid7.client.dcerpc.io.PacketInput;
+import com.rapid7.client.dcerpc.io.PacketOutput;
+import com.rapid7.client.dcerpc.io.ndr.Alignment;
+import com.rapid7.client.dcerpc.io.ndr.Marshallable;
+import com.rapid7.client.dcerpc.io.ndr.Unmarshallable;
+
+/**
+<p>The RPC_SECURITY_DESCRIPTOR structure represents the <a href="https://msdn.microsoft.com/en-us/library/cc244893.aspx#gt_8a7f6700-8311-45bc-af10-82e10accd331">RPC</a> security descriptors.</p>
+<pre> typedef struct _RPC_SECURITY_DESCRIPTOR {
+   [size_is(cbInSecurityDescriptor), length_is(cbOutSecurityDescriptor)]
+     PBYTE lpSecurityDescriptor;
+   DWORD cbInSecurityDescriptor;
+   DWORD cbOutSecurityDescriptor;
+ } RPC_SECURITY_DESCRIPTOR,
+ *  *PRPC_SECURITY_DESCRIPTOR;
+</pre>
+<p><strong>lpSecurityDescriptor:</strong>  A buffer that
+contains a <a href="https://msdn.microsoft.com/en-us/library/cc230366.aspx">SECURITY_DESCRIPTOR</a>,
+as specified in <a href="https://msdn.microsoft.com/en-us/library/cc230273.aspx">[MS-DTYP]</a>
+section 2.4.6.</p>
+<p><strong>cbInSecurityDescriptor:</strong>  The size in
+bytes of the security descriptor.</p>
+<p><strong>cbOutSecurityDescriptor:</strong>  The size
+in bytes of the security descriptor.</p>
+@see <a href="https://msdn.microsoft.com/en-us/library/cc244924.aspx">https://msdn.microsoft.com/en-us/library/cc244924.aspx</a>
+ */
+public class RPCSecurityDescriptor implements Unmarshallable, Marshallable {
+    private int cbInSecurityDescriptor;
+    private int cbOutSecurityDescriptor;
+    private byte[] rawSecurityDescriptor;
+
+    public void setCbInSecurityDescriptor(final int size) {
+        this.cbInSecurityDescriptor = size;
+    }
+
+    public void setCbOutSecurityDescriptor(final int size) {
+        this.cbOutSecurityDescriptor = size;
+    }
+
+    public void setSecurityDescriptor(final byte[] securityDescriptorBuf) {
+        this.rawSecurityDescriptor = securityDescriptorBuf;
+    }
+
+    @Override
+    public void unmarshalPreamble(final PacketInput in) throws IOException {
+    }
+
+    @Override
+    public void unmarshalEntity(final PacketInput in) throws IOException {
+        in.align(Alignment.FOUR);
+        final int refID = in.readReferentID();
+        if (refID != 0) {
+            cbInSecurityDescriptor = in.readInt();
+            cbOutSecurityDescriptor = in.readInt();
+            rawSecurityDescriptor = new byte[cbOutSecurityDescriptor];
+        }
+    }
+
+    @Override
+    public void unmarshalDeferrals(final PacketInput in) throws IOException {
+        if (rawSecurityDescriptor == null)
+            return;
+        in.align(Alignment.FOUR);
+        in.readInt();
+        in.readInt(); // offset
```

Comment:
We should consider the offset here; likely microsoft is not actually using it, but it's possible that they (or some other target) can.
```java
in.fullySkipBytes(4);
final int offset = in.readInt();
final int size = in.readInt();
if (size != rawSecurityDescriptor.length) {
   throw new UnmarshalException(...)
}
// Skip offset (if any)
in.fullySkipBytes(offset);
in.readRawBytes(rawSecurityDescriptor);
```

Reply:
I would be cautious about assuming this is how it would work if there's offset. Given the context of how this struct is used, the offset could only be 0.